{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 19133,
     "status": "ok",
     "timestamp": 1694015835886,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     },
     "user_tz": 240
    },
    "id": "s-9aa8yrHj9j",
    "outputId": "fc7ed900-deb1-4e47-87ba-d43de1e89b8d"
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Mounted at /content/drive\n"
     ]
    }
   ],
   "source": [
    "from google.colab import drive\n",
    "drive.mount('/content/drive')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "executionInfo": {
     "elapsed": 6015,
     "status": "ok",
     "timestamp": 1694015841897,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     },
     "user_tz": 240
    },
    "id": "liJJGzQp4qzO",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "outputId": "2189c9ab-d134-4f22-9ae4-c8e8d350e568"
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Requirement already satisfied: toolz in /usr/local/lib/python3.10/dist-packages (0.12.0)\n",
      "Collecting scikit-allel\n",
      "  Downloading scikit_allel-1.3.6-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (7.4 MB)\n",
      "\u001B[2K     \u001B[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001B[0m \u001B[32m7.4/7.4 MB\u001B[0m \u001B[31m20.8 MB/s\u001B[0m eta \u001B[36m0:00:00\u001B[0m\n",
      "\u001B[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from scikit-allel) (1.23.5)\n",
      "Requirement already satisfied: dask[array] in /usr/local/lib/python3.10/dist-packages (from scikit-allel) (2023.8.1)\n",
      "Requirement already satisfied: click>=8.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (8.1.7)\n",
      "Requirement already satisfied: cloudpickle>=1.5.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (2.2.1)\n",
      "Requirement already satisfied: fsspec>=2021.09.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (2023.6.0)\n",
      "Requirement already satisfied: packaging>=20.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (23.1)\n",
      "Requirement already satisfied: partd>=1.2.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (1.4.0)\n",
      "Requirement already satisfied: pyyaml>=5.3.1 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (6.0.1)\n",
      "Requirement already satisfied: importlib-metadata>=4.13.0 in /usr/local/lib/python3.10/dist-packages (from dask[array]->scikit-allel) (6.8.0)\n",
      "Requirement already satisfied: zipp>=0.5 in /usr/local/lib/python3.10/dist-packages (from importlib-metadata>=4.13.0->dask[array]->scikit-allel) (3.16.2)\n",
      "Requirement already satisfied: locket in /usr/local/lib/python3.10/dist-packages (from partd>=1.2.0->dask[array]->scikit-allel) (1.0.0)\n",
      "Installing collected packages: scikit-allel\n",
      "Successfully installed scikit-allel-1.3.6\n"
     ]
    }
   ],
   "source": [
    "!pip install toolz scikit-allel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "RjGOO5PdFPf7"
   },
   "source": [
    "## Setup"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "id": "odmhCqSVFPf8",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694015845279,
     "user_tz": 240,
     "elapsed": 3385,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "# os.environ[\"MODIN_CPUS\"] = \"8\"\n",
    "# from distributed import Client\n",
    "# client = Client()\n",
    "import glob\n",
    "import numpy as np\n",
    "import math\n",
    "import re\n",
    "import random\n",
    "import shutil\n",
    "import gzip\n",
    "import pandas as pd\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.utils import resample\n",
    "from sklearn import metrics\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from sklearn.metrics import f1_score\n",
    "# import allel\n",
    "from scipy.spatial.distance import squareform\n",
    "from scipy.special import softmax\n",
    "%matplotlib inline\n",
    "from toolz import interleave\n",
    "from tqdm import tqdm\n",
    "from matplotlib import pyplot as plt\n",
    "import tensorflow_datasets as tfds\n",
    "from sklearn.metrics import mean_squared_error\n",
    "from sklearn.linear_model import LassoCV, ElasticNetCV\n",
    "from sklearn.model_selection import KFold,StratifiedKFold"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "A77GFE3xFPf8"
   },
   "source": [
    "## Prepare the data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "id": "PamDEI0R5xMN",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694015845280,
     "user_tz": 240,
     "elapsed": 5,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "outputs": [],
   "source": [
    "class DataLoader:\n",
    "    \"\"\"\n",
    "    If the reference is unphased, cannot handle phased target data, so the valid (ref, target) combinations are:\n",
    "    (phased, phased), (phased, unphased), (unphased, unphased)\n",
    "    Important note: for each case, the model should be trained separately\n",
    "    \"\"\"\n",
    "    def __init__(self, reference_panel_file_path, target_file_path):\n",
    "        self.ref_n_header_lines = []\n",
    "        self.ref_n_data_header = \"\"\n",
    "        self.map_values_1_vec = np.vectorize(self.map_hap_2_ind_parent_1)\n",
    "        self.map_values_2_vec = np.vectorize(self.map_hap_2_ind_parent_2)\n",
    "        print(\"Rading the reference file...\")\n",
    "        # get header\n",
    "        root, ext = os.path.splitext(reference_panel_file_path)\n",
    "        with gzip.open(reference_panel_file_path, 'rt') if ext == '.gz' else open(reference_panel_file_path, 'rt') as f_in:\n",
    "            # skip info\n",
    "            while True:\n",
    "                line = f_in.readline()\n",
    "                if line.startswith(\"##\"):\n",
    "                    self.ref_n_header_lines.append(line)\n",
    "                else:\n",
    "                    self.ref_n_data_header = line\n",
    "                    break\n",
    "        self.reference_panel = pd.read_csv(reference_panel_file_path,\n",
    "                                           comment='#',\n",
    "                                           sep='\\t',\n",
    "                                           names=self.ref_n_data_header.strip().split('\\t'))\n",
    "        self.VARIANT_COUNT = self.reference_panel.shape[0]\n",
    "        print(f\"{self.VARIANT_COUNT} variants found. Done!\")\n",
    "        print(\"Rading the target file...\")\n",
    "        self.target_n_header_lines = []\n",
    "        self.target_n_data_header = \"\"\n",
    "        root, ext = os.path.splitext(target_file_path)\n",
    "        # get header\n",
    "        with gzip.open(target_file_path, 'rt') if ext == '.gz' else open(target_file_path, 'rt') as f_in:\n",
    "            # skip info\n",
    "            while True:\n",
    "                line = f_in.readline()\n",
    "                if line.startswith(\"##\"):\n",
    "                    self.target_n_header_lines.append(line)\n",
    "                else:\n",
    "                    self.target_n_data_header = line\n",
    "                    break\n",
    "        real_target_set = pd.read_csv(target_file_path,\n",
    "                                           comment='#',\n",
    "                                           sep='\\t',\n",
    "                                           names=self.target_n_data_header.strip().split('\\t'),)\n",
    "        print(f\"{real_target_set.shape[0]} variants found. Done!\")\n",
    "        target_is_phased = \"|\" in real_target_set.iloc[0, 10]\n",
    "        ref_is_phased = \"|\" in self.reference_panel.iloc[0, 10]\n",
    "        self.is_phased = target_is_phased and ref_is_phased\n",
    "        print(\"Creating the new target dataframe\")\n",
    "        self.target_set = real_target_set.merge(self.reference_panel[\"ID\"], on='ID', how='right')\n",
    "        self.target_set[self.reference_panel.columns[:9]] = self.reference_panel[self.reference_panel.columns[:9]]\n",
    "        self.target_set.fillna(\".|.\" if self.is_phased else \"./.\", inplace=True)\n",
    "        print(\"Extracting genotype information...\")\n",
    "        SEP = \"|\" if self.is_phased else \"/\"\n",
    "        def get_num_allels(g):\n",
    "            v1, v2 = g.split(SEP)\n",
    "            return max(int(v1), int(v2)) + 1\n",
    "\n",
    "        def key_gen(v1, v2):\n",
    "            return f\"{v1}{SEP}{v2}\"\n",
    "\n",
    "        genotype_vals = np.unique(self.reference_panel.iloc[:, 9:].values)\n",
    "        if target_is_phased != ref_is_phased:\n",
    "            phased_to_unphased_dict = {}\n",
    "            for i in range(genotype_vals.shape[0]):\n",
    "                key = genotype_vals[i]\n",
    "                v1, v2 = [int(s) for s in genotype_vals[i].split(\"|\")]\n",
    "                genotype_vals[i] = f\"{min(v1, v2)}{SEP}{max(v1, v2)}\"\n",
    "                phased_to_unphased_dict[key] = genotype_vals[i]\n",
    "            self.reference_panel.replace(phased_to_unphased_dict, inplace=True)\n",
    "        genotype_vals = np.unique(genotype_vals)\n",
    "        allele_count = max(map(get_num_allels, genotype_vals))\n",
    "        if self.is_phased:\n",
    "            self.hap_map = {str(i): i for i in range(allele_count)}\n",
    "            self.hap_map.update({\".\": allele_count})\n",
    "            self.r_hap_map = {i:k for k, i in self.hap_map.items()}\n",
    "            self.map_preds_2_allele = np.vectorize(lambda x: self.r_hap_map[x])\n",
    "        self.MISSING_VALUE = self.SEQ_DEPTH = allele_count + 1 if self.is_phased else len(genotype_vals) + 1\n",
    "        self.genotype_keys = np.array([key_gen(i,j) for i in range(allele_count) for j in range(allele_count)]) if self.is_phased else genotype_vals\n",
    "        self.genotype_keys = np.hstack([self.genotype_keys, [\".|.\"] if self.is_phased else [\"./.\"]])\n",
    "        self.replacement_dict = {g:i for i,g in enumerate(self.genotype_keys)}\n",
    "        self.reverse_replacement_dict = {i:g for g,i in self.replacement_dict.items()}\n",
    "\n",
    "    def map_hap_2_ind_parent_1(self, x):\n",
    "        return self.hap_map[x.split('|')[0]]\n",
    "\n",
    "    def map_hap_2_ind_parent_2(self, x):\n",
    "        return self.hap_map[x.split('|')[1]]\n",
    "\n",
    "    def __get_forward_data(self, data: pd.DataFrame):\n",
    "        if self.is_phased:\n",
    "            # break it into haplotypes\n",
    "            _x = np.empty((data.shape[1] * 2, data.shape[0]), dtype=np.int32)\n",
    "\n",
    "            _x[0::2] = self.map_values_1_vec(data.values.T)\n",
    "            _x[1::2] = self.map_values_2_vec(data.values.T)\n",
    "            return _x\n",
    "        else:\n",
    "            return data.replace(self.replacement_dict).values.T.astype(np.int32)\n",
    "\n",
    "    def get_ref_set(self, starting_var_index=None, ending_var_index=None):\n",
    "        if starting_var_index>=0 and ending_var_index>=starting_var_index:\n",
    "            return self.__get_forward_data(self.reference_panel.iloc[starting_var_index:ending_var_index, 9:])\n",
    "        else:\n",
    "            print(\"No variant indices provided or indices not valid, using the whole sequence...\")\n",
    "            return self.__get_forward_data(self.reference_panel.iloc[:, 9:])\n",
    "\n",
    "    def get_target_set(self, starting_var_index=None, ending_var_index=None):\n",
    "        if starting_var_index>=0 and ending_var_index>=starting_var_index:\n",
    "            return self.__get_forward_data(self.target_set.iloc[starting_var_index:ending_var_index, 9:])\n",
    "        else:\n",
    "            print(\"No variant indices provided or indices not valid, using the whole sequence...\")\n",
    "            return self.__get_forward_data(self.target_set.iloc[:, 9:])\n",
    "\n",
    "    def convert_haps_to_genotypes(self, allele_probs):\n",
    "      '''output format: GT:DS:GP'''\n",
    "      FORMAT = \"GT:DS:GP\"\n",
    "      n_haploids, n_variants, n_alleles = allele_probs.shape\n",
    "      allele_probs_normalized = softmax(allele_probs, axis=-1)\n",
    "\n",
    "      if n_haploids % 2 != 0:\n",
    "          raise ValueError(\"Number of haploids should be even.\")\n",
    "\n",
    "      n_samples = n_haploids // 2\n",
    "      genotypes = np.zeros((n_samples, n_variants), dtype=object)\n",
    "\n",
    "      for i in tqdm(range(n_samples)):\n",
    "        haploid_1 = allele_probs_normalized[2 * i]\n",
    "        haploid_2 = allele_probs_normalized[2 * i + 1]\n",
    "\n",
    "        for j in range(n_variants):\n",
    "          phased_probs = np.multiply.outer(haploid_1[j], haploid_2[j]).flatten()\n",
    "          unphased_probs = np.array([phased_probs[0], sum(phased_probs[1:3]), phased_probs[-1]])\n",
    "          unphased_probs_str = \",\".join([f\"{v:.6f}\" for v in unphased_probs])\n",
    "          alt_dosage = np.dot(unphased_probs, [0, 1, 2])\n",
    "          variant_genotypes = [str(v) for v in np.argmax(allele_probs_normalized[i*2:(i+1)*2, j], axis=-1)]\n",
    "          genotypes[i, j] = '|'.join(variant_genotypes) + f\":{alt_dosage:.3f}:{unphased_probs_str}\"\n",
    "\n",
    "      new_vcf = self.target_set.copy()\n",
    "      new_vcf.iloc[:n_variants, 9:] = genotypes.T\n",
    "      new_vcf[\"FORMAT\"] = FORMAT\n",
    "      new_vcf[\"QUAL\"] = \".\"\n",
    "      new_vcf[\"FILTER\"] = \".\"\n",
    "      new_vcf[\"INFO\"] = \"IMPUTED\"\n",
    "      return new_vcf\n",
    "\n",
    "    def convert_unphased_probs_to_genotypes(self, allele_probs):\n",
    "      '''output format: GT:DS:GP'''\n",
    "      FORMAT = \"GT:DS:GP\"\n",
    "      n_samples, n_variants, n_alleles = allele_probs.shape\n",
    "      allele_probs_normalized = softmax(allele_probs, axis=-1)\n",
    "      genotypes = np.zeros((n_samples, n_variants), dtype=object)\n",
    "\n",
    "      for i in tqdm(range(n_samples)):\n",
    "          for j in range(n_variants):\n",
    "              unphased_probs = allele_probs_normalized[i, j]\n",
    "              unphased_probs_str = \",\".join([f\"{v:.6f}\" for v in unphased_probs])\n",
    "              alt_dosage = np.dot(unphased_probs, [0, 1, 2])\n",
    "              variant_genotypes = np.vectorize(self.reverse_replacement_dict.get)(np.argmax(unphased_probs, axis=-1)).flatten()\n",
    "              genotypes[i, j] = '/'.join(variant_genotypes) + f\":{unphased_probs_str}:{alt_dosage:.3f}\"\n",
    "\n",
    "      new_vcf = self.target_set.copy()\n",
    "      new_vcf.iloc[:, 9:] = genotypes.T\n",
    "      new_vcf[\"FORMAT\"] = FORMAT\n",
    "      new_vcf[\"QUAL\"] = \".\"\n",
    "      new_vcf[\"FILTER\"] = \".\"\n",
    "      new_vcf[\"INFO\"] = \"IMPUTED\"\n",
    "      return new_vcf\n",
    "\n",
    "    def __get_headers_for_output(self):\n",
    "      headers = [\"##fileformat=VCFv4.2\",\n",
    "           '''##source=STI v1.0.0''',\n",
    "           '''##INFO=<ID=IMPUTED,Number=0,Type=Flag,Description=\"Marker was imputed\">''',\n",
    "           '''##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">''',\n",
    "           '''##FORMAT=<ID=DS,Number=A,Type=Float,Description=\"Estimated Alternate Allele Dosage : [P(0/1)+2*P(1/1)]\">''',\n",
    "           '''##FORMAT=<ID=GP,Number=G,Type=Float,Description=\"Estimated Posterior Probabilities for Genotypes 0/0, 0/1 and 1/1\">''']\n",
    "      return headers\n",
    "\n",
    "    def preds_to_genotypes(self, preds):\n",
    "        \"\"\"\n",
    "        WARNING: This only supports bi-allelic data right now!\n",
    "        :param preds: numpy array of (n_samples, n_variants, n_alleles)\n",
    "        :return: numpy array of the same shape, with genotype calls, e.g., \"0/1\"\n",
    "        \"\"\"\n",
    "        if self.is_phased:\n",
    "          return self.convert_haps_to_genotypes(preds)\n",
    "        else:\n",
    "          return self.convert_unphased_probs_to_genotypes(preds)\n",
    "\n",
    "    def write_ligated_results_to_vcf(self, df, file_name):\n",
    "      with gzip.open(file_name, 'wt') if file_name.endswith(\".gz\") else open(file_name, 'wt') as f_out:\n",
    "          # write info\n",
    "          f_out.write(\"\\n\".join(self.__get_headers_for_output())+\"\\n\")\n",
    "      df.to_csv(file_name, sep=\"\\t\", mode='a', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "id": "j3zy8i_8FPf_",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694015845280,
     "user_tz": 240,
     "elapsed": 5,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "outputs": [],
   "source": [
    "root_data_dir = '[data_path]'\n",
    "train_file_name = \"beadchip_reference_all_minaf_05_snps_hwe_1e-2_filtered_train.vcf.gz\"\n",
    "test_file_name = \"test_data_beadchip_hwe_filtered.vcf.gz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "id": "H2LdnKzXnW0P",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "outputId": "3ef85da6-6102-4979-836d-0157660d93b9",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016015873,
     "user_tz": 240,
     "elapsed": 170597,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Rading the reference file...\n",
      "31143 variants found. Done!\n",
      "Rading the target file...\n",
      "7336 variants found. Done!\n",
      "Creating the new target dataframe\n",
      "Extracting genotype information...\n"
     ]
    }
   ],
   "source": [
    "dl = DataLoader(root_data_dir + train_file_name,\n",
    "                root_data_dir + test_file_name)"
   ]
  },
  {
   "cell_type": "code",
   "source": [
    "np.unique(dl.get_target_set(0, 31143).astype(np.int32))"
   ],
   "metadata": {
    "id": "DuOXiQmlrn9U",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016019306,
     "user_tz": 240,
     "elapsed": 3437,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    },
    "outputId": "b226950c-5731-40e7-f81e-e4ac4c4cd11a"
   },
   "execution_count": 7,
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "array([0, 1, 2], dtype=int32)"
      ]
     },
     "metadata": {},
     "execution_count": 7
    }
   ]
  },
  {
   "cell_type": "code",
   "source": [
    "dl.is_phased"
   ],
   "metadata": {
    "id": "H2d0jNyGLuLB",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016019307,
     "user_tz": 240,
     "elapsed": 11,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    },
    "outputId": "6f32523a-cfce-44f1-9c76-5b1e0cb62128"
   },
   "execution_count": 8,
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "execution_count": 8
    }
   ]
  },
  {
   "cell_type": "code",
   "source": [
    "dl.hap_map"
   ],
   "metadata": {
    "id": "9-Si4oCXbhNZ",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016019308,
     "user_tz": 240,
     "elapsed": 9,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    },
    "outputId": "c8252299-896c-44d9-dcda-7581fbfae907"
   },
   "execution_count": 9,
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "{'0': 0, '1': 1, '.': 2}"
      ]
     },
     "metadata": {},
     "execution_count": 9
    }
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "id": "TOA_NexzN5Qq"
   },
   "source": [
    "## Ligate"
   ]
  },
  {
   "cell_type": "code",
   "source": [
    "assay_name = \"beadchip\"\n",
    "files_prefix = f\"{assay_name}_probs_window\"\n",
    "save_dir = f\"[save_dir]\"\n",
    "output_path = save_dir + f\"STI_{assay_name}_ligated_hmr.vcf.gz\""
   ],
   "metadata": {
    "id": "m2ADpF82jCUf",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016019308,
     "user_tz": 240,
     "elapsed": 7,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "execution_count": 10,
   "outputs": []
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Use this cell if your model is outputting several windows of variants"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "source": [
    "# exp_files_prefix = f\"{assay_name}_exp_probs_window\"\n",
    "chunk_paths = glob.glob(save_dir+files_prefix+\"*\")\n",
    "all_nparrays = []\n",
    "# for cp in tqdm(sorted(chunk_paths)):\n",
    "for i in range(16):\n",
    "  # print(cp)\n",
    "  print(f\"[save_dir]/beadchip_probs_window_{i+1}.npy\")\n",
    "  all_nparrays.append(np.load(f\"[save_dir]/beadchip_probs_window_{i+1}.npy\"))\n",
    "preds = np.hstack(all_nparrays)\n",
    "preds.shape"
   ],
   "metadata": {
    "id": "zJCWuheEH03N",
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "outputId": "ec5d53ce-10f7-484c-b27f-bcdc08825580",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1690114634535,
     "user_tz": 240,
     "elapsed": 231,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "execution_count": null,
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_1.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_2.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_3.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_4.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_5.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_6.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_7.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_8.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_9.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_10.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_11.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_12.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_13.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_14.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_15.npy\n",
      "./drive/MyDrive/ShiLab/beadchip/STI/beadchip_probs_window_16.npy\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "(200, 31143, 2)"
      ]
     },
     "metadata": {},
     "execution_count": 12
    }
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "id": "CFkkfi8QGMzi",
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 461
    },
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016129078,
     "user_tz": 240,
     "elapsed": 107654,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    },
    "outputId": "50fe5dd5-0306-48f4-8767-36e0ae52d76b"
   },
   "outputs": [
    {
     "output_type": "stream",
     "name": "stderr",
     "text": [
      "100%|██████████| 100/100 [01:46<00:00,  1.06s/it]\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "       #CHROM       POS               ID REF ALT QUAL FILTER     INFO  \\\n",
       "0          22  16052986  22:16052986:C:A   C   A    .      .  IMPUTED   \n",
       "1          22  16053444  22:16053444:A:T   A   T    .      .  IMPUTED   \n",
       "2          22  16053791  22:16053791:C:A   C   A    .      .  IMPUTED   \n",
       "3          22  16055942  22:16055942:C:T   C   T    .      .  IMPUTED   \n",
       "4          22  16058758  22:16058758:C:A   C   A    .      .  IMPUTED   \n",
       "...       ...       ...              ...  ..  ..  ...    ...      ...   \n",
       "31138      22  51233182  22:51233182:G:A   G   A    .      .  IMPUTED   \n",
       "31139      22  51233312  22:51233312:A:G   A   G    .      .  IMPUTED   \n",
       "31140      22  51233347  22:51233347:T:C   T   C    .      .  IMPUTED   \n",
       "31141      22  51235979  22:51235979:G:A   G   A    .      .  IMPUTED   \n",
       "31142      22  51237063  22:51237063:T:C   T   C    .      .  IMPUTED   \n",
       "\n",
       "         FORMAT                               NA20534  ...  \\\n",
       "0      GT:DS:GP  0|0:0.590:0.497031,0.415947,0.087023  ...   \n",
       "1      GT:DS:GP  0|0:0.584:0.501467,0.413353,0.085180  ...   \n",
       "2      GT:DS:GP  0|0:0.682:0.434482,0.449341,0.116177  ...   \n",
       "3      GT:DS:GP  1|1:1.318:0.116218,0.449379,0.434404  ...   \n",
       "4      GT:DS:GP  0|0:0.681:0.434878,0.449150,0.115972  ...   \n",
       "...         ...                                   ...  ...   \n",
       "31138  GT:DS:GP  0|0:0.779:0.364262,0.492383,0.143355  ...   \n",
       "31139  GT:DS:GP  0|0:0.700:0.417376,0.465478,0.117147  ...   \n",
       "31140  GT:DS:GP  0|0:0.751:0.380414,0.488556,0.131030  ...   \n",
       "31141  GT:DS:GP  0|0:0.785:0.368889,0.477235,0.153876  ...   \n",
       "31142  GT:DS:GP  0|0:0.729:0.397248,0.476712,0.126040  ...   \n",
       "\n",
       "                                    HG03985  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|1:0.901:0.291665,0.515831,0.192504   \n",
       "31139  0|1:0.854:0.310230,0.525452,0.164318   \n",
       "31140  0|1:0.868:0.302100,0.527790,0.170110   \n",
       "31141  0|0:0.724:0.406733,0.462822,0.130444   \n",
       "31142  0|0:0.713:0.412860,0.461709,0.125430   \n",
       "\n",
       "                                    NA18545  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.618:0.476901,0.428477,0.094622   \n",
       "31139  0|0:0.591:0.495762,0.417147,0.087091   \n",
       "31140  0|0:0.593:0.494266,0.418373,0.087360   \n",
       "31141  0|0:0.651:0.454213,0.440361,0.105426   \n",
       "31142  0|0:0.631:0.468077,0.433339,0.098584   \n",
       "\n",
       "                                    HG01342  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.652:0.454201,0.439491,0.106309   \n",
       "31139  0|0:0.574:0.508013,0.409476,0.082512   \n",
       "31140  0|0:0.579:0.504981,0.411285,0.083734   \n",
       "31141  0|0:0.780:0.372146,0.475791,0.152063   \n",
       "31142  0|0:0.584:0.501487,0.413349,0.085164   \n",
       "\n",
       "                                    HG02154  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.646:0.457918,0.438192,0.103890   \n",
       "31139  0|0:0.571:0.510278,0.408119,0.081603   \n",
       "31140  0|0:0.565:0.514622,0.405506,0.079872   \n",
       "31141  0|0:0.797:0.361684,0.479565,0.158751   \n",
       "31142  0|0:0.578:0.505783,0.410810,0.083406   \n",
       "\n",
       "                                    HG02232  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.680:0.433727,0.452473,0.113799   \n",
       "31139  0|0:0.618:0.476164,0.429209,0.094627   \n",
       "31140  0|0:0.598:0.491172,0.420089,0.088739   \n",
       "31141  0|0:0.803:0.356894,0.483237,0.159869   \n",
       "31142  0|0:0.582:0.502384,0.412927,0.084689   \n",
       "\n",
       "                                    NA19102  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.791:0.359252,0.490259,0.150489   \n",
       "31139  0|0:0.629:0.469392,0.432556,0.098051   \n",
       "31140  0|0:0.682:0.431482,0.454726,0.113792   \n",
       "31141  0|0:0.674:0.439178,0.447564,0.113257   \n",
       "31142  0|0:0.659:0.449065,0.443257,0.107678   \n",
       "\n",
       "                                    NA18526  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.624:0.473219,0.429708,0.097073   \n",
       "31139  0|0:0.562:0.516767,0.404245,0.078989   \n",
       "31140  0|0:0.564:0.515223,0.405224,0.079552   \n",
       "31141  0|0:0.760:0.384311,0.471235,0.144454   \n",
       "31142  0|0:0.573:0.509146,0.408848,0.082006   \n",
       "\n",
       "                                    HG02885  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|1:0.810:0.340886,0.508438,0.150676   \n",
       "31139  0|0:0.660:0.446704,0.446192,0.107104   \n",
       "31140  0|0:0.677:0.435891,0.451631,0.112478   \n",
       "31141  0|0:0.696:0.423546,0.456953,0.119500   \n",
       "31142  0|0:0.634:0.466126,0.433401,0.100474   \n",
       "\n",
       "                                    HG01139  \\\n",
       "0      0|0:0.590:0.497031,0.415947,0.087023   \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180   \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177   \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404   \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972   \n",
       "...                                     ...   \n",
       "31138  0|0:0.742:0.394505,0.469266,0.136229   \n",
       "31139  0|0:0.598:0.491303,0.419259,0.089438   \n",
       "31140  0|0:0.623:0.474274,0.428810,0.096916   \n",
       "31141  0|0:0.663:0.446514,0.443767,0.109720   \n",
       "31142  0|0:0.629:0.470012,0.431247,0.098741   \n",
       "\n",
       "                                    HG03695  \n",
       "0      0|0:0.590:0.497031,0.415947,0.087023  \n",
       "1      0|0:0.584:0.501467,0.413353,0.085180  \n",
       "2      0|0:0.682:0.434482,0.449341,0.116177  \n",
       "3      1|1:1.318:0.116218,0.449379,0.434404  \n",
       "4      0|0:0.681:0.434878,0.449150,0.115972  \n",
       "...                                     ...  \n",
       "31138  0|0:0.756:0.383289,0.477340,0.139371  \n",
       "31139  0|0:0.712:0.411740,0.464736,0.123524  \n",
       "31140  0|0:0.753:0.384101,0.479203,0.136695  \n",
       "31141  0|0:0.719:0.408519,0.463703,0.127778  \n",
       "31142  0|0:0.749:0.388413,0.474174,0.137412  \n",
       "\n",
       "[31143 rows x 109 columns]"
      ],
      "text/html": [
       "\n",
       "  <div id=\"df-24ce9d85-b71e-4845-bdcf-625d54877e05\" class=\"colab-df-container\">\n",
       "    <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>NA20534</th>\n",
       "      <th>...</th>\n",
       "      <th>HG03985</th>\n",
       "      <th>NA18545</th>\n",
       "      <th>HG01342</th>\n",
       "      <th>HG02154</th>\n",
       "      <th>HG02232</th>\n",
       "      <th>NA19102</th>\n",
       "      <th>NA18526</th>\n",
       "      <th>HG02885</th>\n",
       "      <th>HG01139</th>\n",
       "      <th>HG03695</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>22</td>\n",
       "      <td>16052986</td>\n",
       "      <td>22:16052986:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "      <td>0|0:0.590:0.497031,0.415947,0.087023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>22</td>\n",
       "      <td>16053444</td>\n",
       "      <td>22:16053444:A:T</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "      <td>0|0:0.584:0.501467,0.413353,0.085180</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>22</td>\n",
       "      <td>16053791</td>\n",
       "      <td>22:16053791:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "      <td>0|0:0.682:0.434482,0.449341,0.116177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>16055942</td>\n",
       "      <td>22:16055942:C:T</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>...</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "      <td>1|1:1.318:0.116218,0.449379,0.434404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>22</td>\n",
       "      <td>16058758</td>\n",
       "      <td>22:16058758:C:A</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "      <td>0|0:0.681:0.434878,0.449150,0.115972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31138</th>\n",
       "      <td>22</td>\n",
       "      <td>51233182</td>\n",
       "      <td>22:51233182:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.779:0.364262,0.492383,0.143355</td>\n",
       "      <td>...</td>\n",
       "      <td>0|1:0.901:0.291665,0.515831,0.192504</td>\n",
       "      <td>0|0:0.618:0.476901,0.428477,0.094622</td>\n",
       "      <td>0|0:0.652:0.454201,0.439491,0.106309</td>\n",
       "      <td>0|0:0.646:0.457918,0.438192,0.103890</td>\n",
       "      <td>0|0:0.680:0.433727,0.452473,0.113799</td>\n",
       "      <td>0|0:0.791:0.359252,0.490259,0.150489</td>\n",
       "      <td>0|0:0.624:0.473219,0.429708,0.097073</td>\n",
       "      <td>0|1:0.810:0.340886,0.508438,0.150676</td>\n",
       "      <td>0|0:0.742:0.394505,0.469266,0.136229</td>\n",
       "      <td>0|0:0.756:0.383289,0.477340,0.139371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31139</th>\n",
       "      <td>22</td>\n",
       "      <td>51233312</td>\n",
       "      <td>22:51233312:A:G</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.700:0.417376,0.465478,0.117147</td>\n",
       "      <td>...</td>\n",
       "      <td>0|1:0.854:0.310230,0.525452,0.164318</td>\n",
       "      <td>0|0:0.591:0.495762,0.417147,0.087091</td>\n",
       "      <td>0|0:0.574:0.508013,0.409476,0.082512</td>\n",
       "      <td>0|0:0.571:0.510278,0.408119,0.081603</td>\n",
       "      <td>0|0:0.618:0.476164,0.429209,0.094627</td>\n",
       "      <td>0|0:0.629:0.469392,0.432556,0.098051</td>\n",
       "      <td>0|0:0.562:0.516767,0.404245,0.078989</td>\n",
       "      <td>0|0:0.660:0.446704,0.446192,0.107104</td>\n",
       "      <td>0|0:0.598:0.491303,0.419259,0.089438</td>\n",
       "      <td>0|0:0.712:0.411740,0.464736,0.123524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31140</th>\n",
       "      <td>22</td>\n",
       "      <td>51233347</td>\n",
       "      <td>22:51233347:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.751:0.380414,0.488556,0.131030</td>\n",
       "      <td>...</td>\n",
       "      <td>0|1:0.868:0.302100,0.527790,0.170110</td>\n",
       "      <td>0|0:0.593:0.494266,0.418373,0.087360</td>\n",
       "      <td>0|0:0.579:0.504981,0.411285,0.083734</td>\n",
       "      <td>0|0:0.565:0.514622,0.405506,0.079872</td>\n",
       "      <td>0|0:0.598:0.491172,0.420089,0.088739</td>\n",
       "      <td>0|0:0.682:0.431482,0.454726,0.113792</td>\n",
       "      <td>0|0:0.564:0.515223,0.405224,0.079552</td>\n",
       "      <td>0|0:0.677:0.435891,0.451631,0.112478</td>\n",
       "      <td>0|0:0.623:0.474274,0.428810,0.096916</td>\n",
       "      <td>0|0:0.753:0.384101,0.479203,0.136695</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31141</th>\n",
       "      <td>22</td>\n",
       "      <td>51235979</td>\n",
       "      <td>22:51235979:G:A</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.785:0.368889,0.477235,0.153876</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.724:0.406733,0.462822,0.130444</td>\n",
       "      <td>0|0:0.651:0.454213,0.440361,0.105426</td>\n",
       "      <td>0|0:0.780:0.372146,0.475791,0.152063</td>\n",
       "      <td>0|0:0.797:0.361684,0.479565,0.158751</td>\n",
       "      <td>0|0:0.803:0.356894,0.483237,0.159869</td>\n",
       "      <td>0|0:0.674:0.439178,0.447564,0.113257</td>\n",
       "      <td>0|0:0.760:0.384311,0.471235,0.144454</td>\n",
       "      <td>0|0:0.696:0.423546,0.456953,0.119500</td>\n",
       "      <td>0|0:0.663:0.446514,0.443767,0.109720</td>\n",
       "      <td>0|0:0.719:0.408519,0.463703,0.127778</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31142</th>\n",
       "      <td>22</td>\n",
       "      <td>51237063</td>\n",
       "      <td>22:51237063:T:C</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>.</td>\n",
       "      <td>.</td>\n",
       "      <td>IMPUTED</td>\n",
       "      <td>GT:DS:GP</td>\n",
       "      <td>0|0:0.729:0.397248,0.476712,0.126040</td>\n",
       "      <td>...</td>\n",
       "      <td>0|0:0.713:0.412860,0.461709,0.125430</td>\n",
       "      <td>0|0:0.631:0.468077,0.433339,0.098584</td>\n",
       "      <td>0|0:0.584:0.501487,0.413349,0.085164</td>\n",
       "      <td>0|0:0.578:0.505783,0.410810,0.083406</td>\n",
       "      <td>0|0:0.582:0.502384,0.412927,0.084689</td>\n",
       "      <td>0|0:0.659:0.449065,0.443257,0.107678</td>\n",
       "      <td>0|0:0.573:0.509146,0.408848,0.082006</td>\n",
       "      <td>0|0:0.634:0.466126,0.433401,0.100474</td>\n",
       "      <td>0|0:0.629:0.470012,0.431247,0.098741</td>\n",
       "      <td>0|0:0.749:0.388413,0.474174,0.137412</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>31143 rows × 109 columns</p>\n",
       "</div>\n",
       "    <div class=\"colab-df-buttons\">\n",
       "\n",
       "  <div class=\"colab-df-container\">\n",
       "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-24ce9d85-b71e-4845-bdcf-625d54877e05')\"\n",
       "            title=\"Convert this dataframe to an interactive table.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
       "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "\n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    .colab-df-buttons div {\n",
       "      margin-bottom: 4px;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "    <script>\n",
       "      const buttonEl =\n",
       "        document.querySelector('#df-24ce9d85-b71e-4845-bdcf-625d54877e05 button.colab-df-convert');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      async function convertToInteractive(key) {\n",
       "        const element = document.querySelector('#df-24ce9d85-b71e-4845-bdcf-625d54877e05');\n",
       "        const dataTable =\n",
       "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                    [key], {});\n",
       "        if (!dataTable) return;\n",
       "\n",
       "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "          + ' to learn more about interactive tables.';\n",
       "        element.innerHTML = '';\n",
       "        dataTable['output_type'] = 'display_data';\n",
       "        await google.colab.output.renderOutput(dataTable, element);\n",
       "        const docLink = document.createElement('div');\n",
       "        docLink.innerHTML = docLinkHtml;\n",
       "        element.appendChild(docLink);\n",
       "      }\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "\n",
       "<div id=\"df-b7bf2f6c-92be-4990-8ae1-5e42f0d11c56\">\n",
       "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-b7bf2f6c-92be-4990-8ae1-5e42f0d11c56')\"\n",
       "            title=\"Suggest charts.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "     width=\"24px\">\n",
       "    <g>\n",
       "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
       "    </g>\n",
       "</svg>\n",
       "  </button>\n",
       "\n",
       "<style>\n",
       "  .colab-df-quickchart {\n",
       "      --bg-color: #E8F0FE;\n",
       "      --fill-color: #1967D2;\n",
       "      --hover-bg-color: #E2EBFA;\n",
       "      --hover-fill-color: #174EA6;\n",
       "      --disabled-fill-color: #AAA;\n",
       "      --disabled-bg-color: #DDD;\n",
       "  }\n",
       "\n",
       "  [theme=dark] .colab-df-quickchart {\n",
       "      --bg-color: #3B4455;\n",
       "      --fill-color: #D2E3FC;\n",
       "      --hover-bg-color: #434B5C;\n",
       "      --hover-fill-color: #FFFFFF;\n",
       "      --disabled-bg-color: #3B4455;\n",
       "      --disabled-fill-color: #666;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart {\n",
       "    background-color: var(--bg-color);\n",
       "    border: none;\n",
       "    border-radius: 50%;\n",
       "    cursor: pointer;\n",
       "    display: none;\n",
       "    fill: var(--fill-color);\n",
       "    height: 32px;\n",
       "    padding: 0;\n",
       "    width: 32px;\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart:hover {\n",
       "    background-color: var(--hover-bg-color);\n",
       "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "    fill: var(--button-hover-fill-color);\n",
       "  }\n",
       "\n",
       "  .colab-df-quickchart-complete:disabled,\n",
       "  .colab-df-quickchart-complete:disabled:hover {\n",
       "    background-color: var(--disabled-bg-color);\n",
       "    fill: var(--disabled-fill-color);\n",
       "    box-shadow: none;\n",
       "  }\n",
       "\n",
       "  .colab-df-spinner {\n",
       "    border: 2px solid var(--fill-color);\n",
       "    border-color: transparent;\n",
       "    border-bottom-color: var(--fill-color);\n",
       "    animation:\n",
       "      spin 1s steps(1) infinite;\n",
       "  }\n",
       "\n",
       "  @keyframes spin {\n",
       "    0% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "      border-left-color: var(--fill-color);\n",
       "    }\n",
       "    20% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    30% {\n",
       "      border-color: transparent;\n",
       "      border-left-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    40% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-top-color: var(--fill-color);\n",
       "    }\n",
       "    60% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "    }\n",
       "    80% {\n",
       "      border-color: transparent;\n",
       "      border-right-color: var(--fill-color);\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "    90% {\n",
       "      border-color: transparent;\n",
       "      border-bottom-color: var(--fill-color);\n",
       "    }\n",
       "  }\n",
       "</style>\n",
       "\n",
       "  <script>\n",
       "    async function quickchart(key) {\n",
       "      const quickchartButtonEl =\n",
       "        document.querySelector('#' + key + ' button');\n",
       "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
       "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
       "      try {\n",
       "        const charts = await google.colab.kernel.invokeFunction(\n",
       "            'suggestCharts', [key], {});\n",
       "      } catch (error) {\n",
       "        console.error('Error during call to suggestCharts:', error);\n",
       "      }\n",
       "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
       "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
       "    }\n",
       "    (() => {\n",
       "      let quickchartButtonEl =\n",
       "        document.querySelector('#df-b7bf2f6c-92be-4990-8ae1-5e42f0d11c56 button');\n",
       "      quickchartButtonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "    })();\n",
       "  </script>\n",
       "</div>\n",
       "\n",
       "  <div id=\"id_eedf8293-6ff0-479a-b15b-da8b43004a97\">\n",
       "    <style>\n",
       "      .colab-df-generate {\n",
       "        background-color: #E8F0FE;\n",
       "        border: none;\n",
       "        border-radius: 50%;\n",
       "        cursor: pointer;\n",
       "        display: none;\n",
       "        fill: #1967D2;\n",
       "        height: 32px;\n",
       "        padding: 0 0 0 0;\n",
       "        width: 32px;\n",
       "      }\n",
       "\n",
       "      .colab-df-generate:hover {\n",
       "        background-color: #E2EBFA;\n",
       "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "        fill: #174EA6;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate {\n",
       "        background-color: #3B4455;\n",
       "        fill: #D2E3FC;\n",
       "      }\n",
       "\n",
       "      [theme=dark] .colab-df-generate:hover {\n",
       "        background-color: #434B5C;\n",
       "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "        fill: #FFFFFF;\n",
       "      }\n",
       "    </style>\n",
       "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('genotypes')\"\n",
       "            title=\"Generate code using this dataframe.\"\n",
       "            style=\"display:none;\">\n",
       "\n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
       "  </svg>\n",
       "    </button>\n",
       "    <script>\n",
       "      (() => {\n",
       "      const buttonEl =\n",
       "        document.querySelector('#id_eedf8293-6ff0-479a-b15b-da8b43004a97 button.colab-df-generate');\n",
       "      buttonEl.style.display =\n",
       "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "      buttonEl.onclick = () => {\n",
       "        google.colab.notebook.generateWithVariable('genotypes');\n",
       "      }\n",
       "      })();\n",
       "    </script>\n",
       "  </div>\n",
       "\n",
       "    </div>\n",
       "  </div>\n"
      ]
     },
     "metadata": {},
     "execution_count": 12
    }
   ],
   "source": [
    "genotypes = dl.preds_to_genotypes(preds)\n",
    "genotypes"
   ]
  },
  {
   "cell_type": "code",
   "source": [
    "dl.write_ligated_results_to_vcf(genotypes, output_path)"
   ],
   "metadata": {
    "id": "AZEE3cQItlsa",
    "executionInfo": {
     "status": "ok",
     "timestamp": 1694016170261,
     "user_tz": 240,
     "elapsed": 41191,
     "user": {
      "displayName": "net crowmaster",
      "userId": "14126946344391667924"
     }
    }
   },
   "execution_count": 13,
   "outputs": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "xQ7TbkVjCnw8"
   },
   "outputs": [],
   "source": [
    "from google.colab import runtime\n",
    "runtime.unassign()"
   ]
  },
  {
   "cell_type": "code",
   "source": [],
   "metadata": {
    "id": "qRXjOek2d-Ok"
   },
   "execution_count": null,
   "outputs": []
  }
 ],
 "metadata": {
  "colab": {
   "provenance": [
    {
     "file_id": "1lo9cmptjRRESsVlVECvlJFHbLeaygq2s",
     "timestamp": 1640128751823
    },
    {
     "file_id": "https://github.com/keras-team/keras-io/blob/master/examples/vision/ipynb/perceiver_image_classification.ipynb",
     "timestamp": 1621552889682
    }
   ]
  },
  "kernelspec": {
   "display_name": "Python 3",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
